turbopack/crates/turbo-tasks-fs/src/lib.rs

```diff
@@ -307,25 +303,31 @@ impl DiskFileSystem {
         });
     }
 
-    pub async fn start_watching(&self, poll_interval: Option<Duration>) -> Result<()> {
-        self.start_watching_internal(false, poll_interval).await
-    }
-
-    pub async fn start_watching_with_invalidation_reason(
-        &self,
-        poll_interval: Option<Duration>,
-    ) -> Result<()> {
-        self.start_watching_internal(true, poll_interval).await
+    fn invalidate_from_write(&self, full_path: &Path, invalidators: HashSet<Invalidator>) {
+        if !invalidators.is_empty() {
+            if let Some(path) = format_absolute_fs_path(full_path, &self.name, self.root_path()) {
+                if invalidators.len() == 1 {
+                    let invalidator = invalidators.into_iter().next().unwrap();
+                    invalidator.invalidate_with_reason(Write { path });
+                } else {
```

Comment:
I'm a bit skeptical that this is a hot enough codepath for this special-casing to be worth it, but a better way to generalize this optimization would be:
- Always save the first invalidator into an `Option<Invalidator>`.
- Loop over all remaining invalidators (with `for_each` or a normal `for` loop) using `path.clone()`.
- After the loop, `if let Some(invalidator) = first_invalidator`, call `invalidate_with_reason` using the owned path.